File: package/tea-cup-link-pagination/app/example-app-backend/src/Domain/Type/Room.hs
{-# LANGUAGE DeriveAnyClass #-}
{-# LANGUAGE DeriveGeneric #-}

module Domain.Type.Room
  ( RoomDTO (..)
  , LastMessagePreviewDTO (..)
  ) where

import Data.Aeson (FromJSON, ToJSON)
import Data.Int (Int64)
import Data.Text (Text)
import GHC.Generics (Generic)

data LastMessagePreviewDTO = LastMessagePreviewDTO
  { id :: Text
  , authorName :: Text
  , content :: Text
  , timestamp :: Int64
  }
  deriving (Show, Eq, Generic, ToJSON, FromJSON)

data RoomDTO = RoomDTO
  { id :: Text
  , name :: Text
  , topic :: Text
  , icon :: Text
  , unreadCount :: Int
  , firstUnreadChatId :: Maybe Text
  , orderIndex :: Int
  , membersCount :: Int
  , isPrivate :: Bool
  , lastMessage :: Maybe LastMessagePreviewDTO
  , lastUnreadMessage :: Maybe LastMessagePreviewDTO
  }
  deriving (Show, Eq, Generic, ToJSON, FromJSON)

File: package/tea-cup-link-pagination/app/example-app/src/api/type/room.ts
import * as EqClass from 'fp-ts/lib/Eq'
import * as OrdClass from 'fp-ts/lib/Ord'
import { pipe } from 'fp-ts/lib/function'
import * as N from 'fp-ts/lib/number'
import * as S from 'fp-ts/lib/string'

import { type Pagination } from './pagination'

export type LastMessagePreview = {
  readonly id: string
  readonly authorName: string
  readonly content: string
  readonly timestamp: number
}

export type Room = {
  id: string
  name: string
  topic: string
  icon: string
  unreadCount: number
  firstUnreadChatId?: string | null
  orderIndex: number
  membersCount: number
  isPrivate?: boolean
  lastMessage?: LastMessagePreview | null
  lastUnreadMessage?: LastMessagePreview | null
}

export const RoomEq: EqClass.Eq<Room> = EqClass.struct<Room>({
  id: S.Eq,
  name: S.Eq,
  topic: S.Eq,
  icon: S.Eq,
  unreadCount: N.Eq,
  firstUnreadChatId: {
    equals: (a?: string | null, b?: string | null) => a === b,
  },
  orderIndex: N.Eq,
  membersCount: N.Eq,
  isPrivate: {
    equals: (a?: boolean, b?: boolean) => a === b,
  },
  lastMessage: {
    equals: (a?: LastMessagePreview | null, b?: LastMessagePreview | null) =>
      a?.id === b?.id &&
      a?.content === b?.content &&
      a?.timestamp === b?.timestamp,
  },
  lastUnreadMessage: {
    equals: (a?: LastMessagePreview | null, b?: LastMessagePreview | null) =>
      a?.id === b?.id &&
      a?.content === b?.content &&
      a?.timestamp === b?.timestamp,
  },
})

export const RoomOrd: OrdClass.Ord<Room> = pipe(
  N.Ord,
  OrdClass.contramap((r: Room) => r.lastMessage?.timestamp ?? 0),
  OrdClass.reverse,
)

export type RoomCurrentPrevNextResult = [
  Room,
  Pagination<Room>, // prev
  Pagination<Room>, // next
]

export type GetRoomsParams = {
  targetRoomId?: string | null
  beforeTimestamp?: number
  afterTimestamp?: number
  pageSize?: number
  searchString?: string
  latencyMs?: number
  networkOnline?: boolean
}
